src/track/beatutils.cpp

```diff
@@ -14,453 +14,355 @@
+    double centerBpm = 60 * sampleRate / longesRegionBeatLenth;
 
-        //qDebug() << "Local BPM between beat " << (i-N) << " and " << i << " is " << avg_bpm;
+    qDebug() << "minRoundBpm" << minRoundBpm;
+    qDebug() << "maxRoundBpm" << maxRoundBpm;
+    double roundBpm = roundBpmWithinRange(minRoundBpm, centerBpm, maxRoundBpm);
 
-        // If the local BPM is within kCorrectBeatLocalBpmEpsilon of the global
-        // BPM then use this window as the first beat.
-        if (fabs(global_bpm - avg_bpm) <= kCorrectBeatLocalBpmEpsilon) {
-            //qDebug() << "Using beat " << (i-N) << " as first beat";
-            return start_sample;
-        }
+    if (pFirstBeat) {
+        *pFirstBeat = constantRegions[startRegion].firstBeat;
     }
-
-    // If we didn't find any beat that matched the window, return the first
-    // beat.
-    return !rawbeats.empty() ? rawbeats.first() : 0.0;
+    return roundBpm;
 }
 
 // static
-double BeatUtils::calculateFixedTempoFirstBeat(
-        bool enableOffsetCorrection,
-        const QVector<double>& rawbeats,
-        const int sampleRate,
-        const int totalSamples,
-        const double globalBpm) {
-    if (rawbeats.size() == 0) {
-        return 0;
+double BeatUtils::roundBpmWithinRange(double minBpm, double centerBpm, double maxBpm) {
+    // First try to snap to a full integer BPM
+    double snapBpm = round(centerBpm);
+    if (snapBpm > minBpm && snapBpm < maxBpm) {
+        // Success
+        return snapBpm;
     }
 
-    if (!enableOffsetCorrection) {
-        return rawbeats.first();
+    // Probe the reasonable multipliers for 0.5
+    double roundBpmWidth = maxBpm - minBpm;
+    if (roundBpmWidth > 1.0 / 2) {
+        // 0.5 BPM are only reasonable if the double value is not insane
+        // or the 2/3 value is not too small.
+        if (centerBpm < 85) {
+            // this cane be actually up to 175 BPM
+            // allow halve BPM values
+            return round(centerBpm * 2) / 2;
+        } else if (centerBpm > 127) {
+            // optimize for 2/3 going down to 85
+            return round(centerBpm / 3 * 2) * 3 / 2;
+        }
     }
 
-    QVector <double> corrbeats;
-    // Length of a beat at globalBpm in mono samples.
-    const double beat_length = 60.0 * sampleRate / globalBpm;
-
-
-    double firstCorrectBeat = findFirstCorrectBeat(
-        rawbeats, sampleRate, globalBpm);
-
-    // We start building a fixed beat grid at globalBpm and the first beat from
-    // rawbeats that matches globalBpm.
-    double i = firstCorrectBeat;
-    while (i <= totalSamples) {
-        corrbeats << i;
-        i += beat_length;
+    if (roundBpmWidth > 1.0 / 12) {
+        // this covers all sorts of 1/2 2/3 and 3/4 multiplier
+        return round(centerBpm * 12) / 12;
+    } else {
+        // We are here if we have more that ~75 beats and ~30 s
+        // try to snap to a 1/12 Bpm
+        double snapBpm = round(centerBpm * 12) / 12;
```

Comment:
This shadows the `snapBpm` variable above. Maybe reuse that variable or use another variable name? The alternative would be to wrap the first occurrence of `snapBpm` in braces to define the scope of the variable.